babel/dates.py

```diff
@@ -1161,7 +1165,12 @@ def parse_date(string, locale=LC_TIME, format='medium'):
     if not numbers:
         raise ParseError("No numbers were found in input")
 
-    # TODO: try ISO format first?
+    # we try ISO format first YYYY-MM-DD
+    if len(numbers) == 3:
+        ints = list(map(int, numbers))
+        if ints[0] > 31 and ints[1] <= 12 and ints[2] <= 31:
+            return date(*ints)
```

Comment:
I think this should be wrapped in a try-expect. If the atoms aren't interpretible as integers, this would raise an exception. Additionally since the format is `YYYY`, shouldn't `ints[0] > 31` be `ints[0] >= 1000`? At that point you could also do something like

```
if ints[0] >= 1000 and 1 <= ints[1] <= 12 and 1 <= ints[2] <= 31
```
for better validation.

Reply:
I think, the issue is that we're not aligned on the exact purpose of the code: I was trying to identify everything looking "ISO-ish" without trying to be aligned exactly with the specs. In this perspective, I still think that my code is correct (ISO-ish being a year followed by a month followed by a day). So let me know if I rather should aim at ISO-strictness, where indeed my code would need to look very different.
PS: all my comments will appear at once because I wrongly started a review...